pkg/applicationserver/registry.go

```diff
@@ -33,6 +34,125 @@ type DeviceRegistry interface {
 	Set(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, paths []string, f func(*ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error)) (*ttnpb.EndDevice, error)
 }
 
+type deprecatedDeviceField struct {
+	Old          string
+	New          string
+	GetTransform func(dev *ttnpb.EndDevice)
+	SetTransform func(dev *ttnpb.EndDevice, useOld, useNew bool) error
+}
+
+type deprecatedDeviceFieldMatch struct {
+	deprecatedDeviceField
+	MatchedOld bool
+	MatchedNew bool
+}
+
+type deprecatedDeviceFieldRegistryWrapper struct {
+	fields   []deprecatedDeviceField
+	registry DeviceRegistry
+}
+
+func matchDeprecatedDeviceFields(paths []string, deprecated []deprecatedDeviceField) ([]string, []deprecatedDeviceFieldMatch) {
+	var matched []deprecatedDeviceFieldMatch
+	for _, f := range deprecated {
+		hasOld, hasNew := ttnpb.HasAnyField(paths, f.Old), ttnpb.HasAnyField(paths, f.New)
+		switch {
+		case !hasOld && !hasNew:
+			continue
+		case hasOld && hasNew:
+		case hasOld:
+			paths = ttnpb.AddFields(paths, f.New)
+		case hasNew:
+			paths = ttnpb.AddFields(paths, f.Old)
+		}
+		matched = append(matched, deprecatedDeviceFieldMatch{
+			deprecatedDeviceField: f,
+			MatchedOld:            hasOld,
+			MatchedNew:            hasNew,
+		})
+	}
+	return paths, matched
+}
+
+func (w deprecatedDeviceFieldRegistryWrapper) Get(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, paths []string) (*ttnpb.EndDevice, error) {
+	paths, deprecated := matchDeprecatedDeviceFields(paths, w.fields)
+	dev, err := w.registry.Get(ctx, ids, paths)
+	if err != nil || dev == nil {
+		return dev, err
+	}
+	for _, d := range deprecated {
+		d.GetTransform(dev)
+	}
+	return dev, nil
+}
+
+func (w deprecatedDeviceFieldRegistryWrapper) Set(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, paths []string, f func(*ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error)) (*ttnpb.EndDevice, error) {
+	paths, deprecated := matchDeprecatedDeviceFields(paths, w.fields)
+	dev, err := w.registry.Set(ctx, ids, paths, func(dev *ttnpb.EndDevice) (*ttnpb.EndDevice, []string, error) {
+		if dev != nil {
+			for _, d := range deprecated {
+				d.GetTransform(dev)
+			}
+		}
+		dev, paths, err := f(dev)
+		if err != nil || dev == nil {
+			return dev, paths, err
+		}
+		for _, d := range deprecated {
+			if ttnpb.HasAnyField(paths, d.Old) {
+				paths = ttnpb.AddFields(paths, d.New)
+			}
```

Comment:
@rvolosatovs is there a reason this isn't in NS?

If a caller sets the old field only, this makes sure the new field is set as well, after the transform ran.

Reply:
Added this to NS registry as well

Reply:
Please remove, see NS comment.

Reply:
It's not; `paths` is set by `f()` (L63), which can be unrelated to the output of `MatchReplacedEndDeviceFields()` (L56)

Reply:
Ah, right. Yeah, indeed that's not the case.
Still, this should not be there - "old" fields should not be used by implementations of servers, only by the clients. The path set returned here should always be either completely generated by the server or processed from user input by the server.

Reply:
Yeah, you're right.